contents/algorithms/graph.md

```diff
@@ -27,29 +27,79 @@ A tree-like diagram could very well be a graph that allows for cycles and a naiv
 - **Uncommon** - Topological Sort, Dijkstra's algorithm
 - **Rare** - Bellman-Ford algorithm, Floyd-Warshall algorithm, Prim's algorithm, Kruskal's algorithm
 
-In coding interviews, graphs are commonly represented as 2-D matrices where cells are the nodes and each cell can traverse to its adjacent cells (up/down/left/right). Hence it is important that you be familiar with traversing a 2-D matrix. When recursively traversing the matrix, always ensure that your next position is within the boundary of the matrix. More tips for doing depth-first searches on a matrix can be found [here](https://discuss.leetcode.com/topic/66065/python-dfs-bests-85-tips-for-all-dfs-in-matrix-question/). A simple template for doing depth-first searches on a matrix goes like this:
+In coding interviews, graphs are commonly represented as 2-D matrices where cells are the nodes and each cell can traverse to its adjacent cells (up/down/left/right). Hence it is important that you be familiar with traversing a 2-D matrix. When traversing the matrix, always ensure that your current position is within the boundary of the matrix and has not been visited before. 
+
+A simple template for doing depth-first searches on a matrix goes like this:
 
 ```py
-def traverse(matrix):
+def dfs(matrix, method):
   rows, cols = len(matrix), len(matrix[0])
   visited = set()
   directions = ((0, 1), (0, -1), (1, 0), (-1, 0))
-  def dfs(i, j):
-    if (i, j) in visited:
+
+  # Depends upon the question: many grid questions have blocked cells.
+  # This implementation assumes 0s represent valid and 1s represent invalid
+  def is_valid(i, j):
```

Comment:
Don't add is_valid here since it only applies to very specific questions. Don't change the original dfs implementation except for the naming.